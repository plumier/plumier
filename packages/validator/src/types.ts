export type Class<T = any> = new (...args: any[]) => T
