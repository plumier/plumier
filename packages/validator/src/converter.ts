import { Class } from "./types"

// --------------------------------------------------------------------- //
// ------------------------------- TYPES ------------------------------- //
// --------------------------------------------------------------------- //

type Converter = (value: {}) => {} | undefined

// --------------------------------------------------------------------- //
// ------------------------------- HELPER ------------------------------ //
// --------------------------------------------------------------------- //

//some object can't simply convertible to string https://github.com/emberjs/ember.js/issues/14922#issuecomment-278986178
function safeToString(value: any) {
    try {
        return value.toString()
    } catch (e) {
        return "[object Object]"
    }
}


// --------------------------------------------------------------------- //
// ----------------------------- CONVERTERS ---------------------------- //
// --------------------------------------------------------------------- //

function booleanConverter(rawValue: {}) {
    const value = safeToString(rawValue)
    const list: { [key: string]: boolean | undefined } = {
        on: true, true: true, "1": true, yes: true,
        off: false, false: false, "0": false, no: false
    }
    return list[value.toLowerCase()]
}

function numberConverter(rawValue: {}) {
    const value = safeToString(rawValue)
    if (value === "") return
    const result = Number(value)
    if (isNaN(result)) return
    return result
}

function dateConverter(rawValue: {}) {
    const value = safeToString(rawValue)
    const stamp = Date.parse(value)
    return isNaN(stamp) ? undefined : new Date(stamp)
}

function stringConverter(rawValue: {}) {
    return safeToString(rawValue)
}

const defaultConverters = new Map<Class, Converter>([
    [Boolean, booleanConverter],
    [Date, dateConverter],
    [Number, numberConverter],
    [String, stringConverter]
])

export { defaultConverters, Converter, safeToString }
