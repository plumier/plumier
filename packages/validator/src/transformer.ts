import reflect, { useCache } from "@plumier/reflect"

import { Class } from "./types"
import { Converter, defaultConverters } from './converter';

// --------------------------------------------------------------------- //
// ------------------------------- TYPES ------------------------------- //
// --------------------------------------------------------------------- //

export interface PrimitiveNode {
    kind: "Primitive"
    type: Class
    converter: Converter
}

export interface ObjectNode {
    kind: "Object",
    type: Class
    properties: Map<string, SuperNode>
}

export interface ArrayNode {
    kind: "Array",
    type: Class
    element: SuperNode
}

export type SuperNode = PrimitiveNode | ObjectNode | ArrayNode

// --------------------------------------------------------------------- //
// -------------------------------- AST -------------------------------- //
// --------------------------------------------------------------------- //

function primitiveNode(type: Class): PrimitiveNode {
    return {
        kind: "Primitive", type,
        converter: defaultConverters.get(type)!
    }
}

function arrayNode(type: Class[]): ArrayNode {
    let defer: SuperNode
    return {
        kind: "Array", type: type[0],
        get element() {
            return defer || (defer = transform(type[0]))
        }
    }
}

function objectNode(type: Class): ObjectNode {
    let defer: Map<string, SuperNode>
    return {
        kind: "Object", type,
        get properties() {
            if (defer)
                return defer
            else {
                const meta = reflect(type)
                return defer = new Map(meta.properties.map(x => (<[string, ObjectNode]>[x.name, transformer(x.type)])))
            }
        }
    }
}

// --------------------------------------------------------------------- //
// ---------------------------- TRANSFORMER ---------------------------- //
// --------------------------------------------------------------------- //

function transformer(type: Class | Class[]) {
    if (type === Boolean || type === Number || type === Date || type === String)
        return primitiveNode(type)
    else if (Array.isArray(type))
        return arrayNode(type)
    else
        return objectNode(type)
}

// --------------------------------------------------------------------- //
// ------------------------------- CACHE ------------------------------- //
// --------------------------------------------------------------------- //

const astCache = new Map<Class, SuperNode>()
const arrayAstCache = new Map<Class, SuperNode>()

const transformArray = useCache(arrayAstCache, transformer, (type: any) => type[0])
const transformAll = useCache(astCache, transformer, (type) => type)

function transform(type: Class | Class[]) {
    const cachedTransformer = Array.isArray(type) ? transformArray : transformAll
    return cachedTransformer(type)
}

export { transform }