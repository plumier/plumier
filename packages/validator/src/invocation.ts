import { Result, ParentInfo } from "./visitor";
import { SuperNode } from './transformer';
import { Class } from './types';


type VisitorExtension = (next: VisitorInvocation) => Result


interface VisitorInvocation {
    value: {},
    type: Class,
    path: string,
    decorators: any[]
    parent?: ParentInfo
    proceed(): Result
}

class ExtensionInvocationImpl implements VisitorInvocation {
    value: {}
    type: Class
    path: string
    decorators: any[]
    parent?:ParentInfo

    constructor(public ext: VisitorExtension, private next: VisitorInvocation) {
        this.value = next.value
        this.type = next.type
        this.path = next.path
        this.decorators = next.decorators
        this.parent = next.parent
    }

    proceed() {
        return this.ext(this.next)
    }
}

class VisitorInvocationImpl implements VisitorInvocation {
    value: {}
    type: Class
    path: string
    decorators: any[]
    parent?:ParentInfo
    constructor(value: {}, path: string, private ast: SuperNode, decorators: any[], private visitor: () => Result, parent?:ParentInfo) {
        this.value = value
        this.type = ast.type
        this.path = path
        this.decorators = decorators
        this.parent = parent
    }

    proceed(): Result {
        return this.visitor()
    }
}

function pipe(value: {}, strPath: string, ast: SuperNode, decorators: any[], extensions: VisitorExtension[], visitor: () => Result, parent?:ParentInfo) {
    return extensions.reduce((prev, cur) => new ExtensionInvocationImpl(cur, prev),
        <VisitorInvocation>new VisitorInvocationImpl(value, strPath, ast, decorators, visitor, parent))
}

export { pipe, VisitorExtension, VisitorInvocation }