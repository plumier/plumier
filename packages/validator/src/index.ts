import { Opt, val } from "./decorators"
import { VisitorExtension, VisitorInvocation } from "./invocation"
import { transform } from "./transformer"
import { Class } from "./types"
import {
    partial,
    PartialValidator,
    Validator,
    validatorVisitor,
    ValidatorDecorator,
    RequiredValidator,
} from "./validation"
import { pipeline, Result, ResultMessages } from "./visitor"
import {defaultConverters} from "./converter"



interface Option<T = any> {
    /**
     * Expected type
     */
    type: T,

    /**
     * List of visitor extension to extend TypedConverter internal process
     */
    visitors?: VisitorExtension[]

    /**
     * Top decorators of the type
     */
    decorators?: any[],

    /**
     * Root path
     */
    path?: string,

    /**
     * Guess single value as array element if expected type is array. Default false
     */
    guessArrayElement?: boolean
}

/**
 * Convert value into type specified on configuration.
 * @param value 
 * @param opt 
 */
function convert<T = any>(value: any, option: Option<T> | T): T extends Class<infer R>[] ? Result<R[]> : T extends Class<infer R> ? Result<R> : Result<any> {
    const opt = "type" in option ? option : { type: option }
    return pipeline(value, transform(opt.type as any), {
        path: opt.path || "", extension: opt.visitors || [],
        decorators: opt.decorators || [],
        guessArrayElement: opt.guessArrayElement || false
    }) as any
}

/**
 * Create type converter with specific configuration
 * @param option 
 */
export default function createConverter<T = any>(option: Option<T> | T) {
    return (value: any) => convert<T>(value, option)
}

function createValidator<T>(option: T | Option<T>) {
    return (value: any) => validate<T>(value, option)
}

function validate<T>(value: any, option: Option<T> | T) {
    const opt = "type" in option ? option : { type: option }
    const visitors = [validatorVisitor]
    for (const visitor of (opt.visitors || [])) {
        visitors.push(visitor)
    }
    return convert(value, {
        decorators: opt.decorators, guessArrayElement: opt.guessArrayElement,
        path: opt.path, type: opt.type, visitors
    })
}

export {
    convert, Option, VisitorExtension,
    VisitorInvocation, Result, ResultMessages,
    partial, validatorVisitor, Validator,
    PartialValidator, val, Opt, validate, createValidator, ValidatorDecorator,
    RequiredValidator, defaultConverters
}