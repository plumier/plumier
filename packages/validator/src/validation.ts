import reflect, { decorateProperty, decorate, mergeDecorator, TypeOverride } from "@plumier/reflect";
import { VisitorInvocation } from './invocation';
import { Result } from './visitor';
import { Class } from './types';
import { safeToString } from './converter';

const RequiredValidator = Symbol("tc:required")
const PartialValidator = Symbol("tc:partial")
type Validator = (val: any) => string | undefined
interface ValidatorDecorator { type: "tc:validator", validator: Validator | string | symbol }

// --------------------------------------------------------------------- //
// ----------------------------- DECORATORS ---------------------------- //
// --------------------------------------------------------------------- //

function createValidation(validator: Validator | string | symbol) {
    return decorateProperty(<ValidatorDecorator>{ type: "tc:validator", validator })
}

function required() {
    return createValidation(RequiredValidator)
}

function partial(type: TypeOverride | ((x: any) => TypeOverride)) {
    return mergeDecorator(reflect.type(type), createValidation(PartialValidator))
}

// --------------------------------------------------------------------- //
// ------------------------------ HELPERS ------------------------------ //
// --------------------------------------------------------------------- //

function isPartial(i: VisitorInvocation) {
    return i.parent && i.parent.decorators.find((x: ValidatorDecorator) => x.validator === PartialValidator)
}

function isEmpty(result: Result) {
    return (result.value === "" || result.value === undefined || result.value === null)
}

// --------------------------------------------------------------------- //
// ------------------------------ VISITORS ----------------------------- //
// --------------------------------------------------------------------- //

function getValidators(i: VisitorInvocation): ValidatorDecorator[] {
    return i.decorators
        .filter((x: ValidatorDecorator) => x.validator !== PartialValidator && x.type === "tc:validator")
}

function validatorVisitor(i: VisitorInvocation): Result {
    const validators = getValidators(i)
    const result = i.proceed()
    if (!isPartial(i) && isEmpty(result) && validators.some(x => x.validator === RequiredValidator))
        return Result.error(i.value, i.path, "Required")
    if (isEmpty(result) || validators.length === 0) return result
    else {
        const messages: string[] = []
        for (const validator of validators) {
            if (typeof validator.validator === "function") {
                const msg = validator.validator(safeToString(i.value))
                if (msg) messages.push(msg)
            }
        }
        return messages.length > 0 ? Result.error(i.value, i.path, messages) : result
    }
}

export {
    createValidation, required, partial,
    validatorVisitor, Validator, ValidatorDecorator,
    PartialValidator, RequiredValidator
}
