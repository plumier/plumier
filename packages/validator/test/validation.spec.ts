import reflect, { decorateMethod } from "@plumier/reflect"

import { PartialValidator, Result, val, validate, ValidatorDecorator, VisitorInvocation } from "../src"
import { RequiredValidator } from "../src/validation"

const required = <ValidatorDecorator>{ type: "tc:validator", validator: RequiredValidator }


describe("Required Validator and Partial", () => {
    describe("Decorators", () => {
        it("Should able to decorate required validator", () => {
            @reflect.parameterProperties()
            class AnimalClass {
                constructor(
                    public id: number,
                    public name: string,
                    public deceased: boolean,
                    @val.required()
                    public birthday: Date
                ) { }
            }
            expect(reflect(AnimalClass)).toMatchSnapshot()
        })

        it("Should able to decorate partial validator", () => {
            @reflect.parameterProperties()
            class AnimalClass {
                constructor(
                    public id: number,
                    public name: string,
                    public deceased: boolean,
                    public birthday: Date
                ) { }
            }

            class AnimalController {
                @decorateMethod({})
                save(@val.partial(AnimalClass) data: AnimalClass) { }
            }
            expect(reflect(AnimalController)).toMatchSnapshot()
        })
    })

    describe("Primitive Type Validation", () => {

        it("Should validate undefined if required", () => {
            const result = validate(undefined, {
                path: "data",
                type: Number,
                decorators: [required]
            })
            expect(result).toMatchSnapshot()
        })
        it("Should validate null if required", () => {
            const result = validate(null, {
                path: "data",
                type: Number,
                decorators: [required]
            })
            expect(result).toMatchSnapshot()
        })
        it("Should validate empty string if required", () => {
            const result = validate("", {
                path: "data",
                type: Number,
                decorators: [required]
            })
            expect(result).toMatchSnapshot()
        })
        it("Should give proper info if provided wrong type of value", () => {
            const result = validate("hula", {
                path: "data",
                type: Number,
            })
            expect(result).toMatchSnapshot()
        })
        it("Should valid if not required and provided null", () => {
            const result = validate(null, {
                path: "data",
                type: Number,
            })
            expect(result).toMatchSnapshot()
        })

        it("Should valid if not required and provided undefined", () => {
            const result = validate(undefined, {
                path: "data",
                type: Number,
            })
            expect(result).toMatchSnapshot()
        })

        it("Should valid if not required and provided empty string", () => {
            const result = validate("", {
                path: "data",
                type: Number,
            })
            expect(result).toMatchSnapshot()
        })
    })

    describe("Object Validator", () => {
        @reflect.parameterProperties()
        class AnimalClass {
            constructor(
                @val.required()
                public id: number,
                public name: string,
                public deceased: boolean,
                public birthday: Date
            ) { }
        }

        const option = { type: AnimalClass, }

        it("Should validate undefined property if required", () => {
            const result = validate({ id: undefined, name: "Mimi", deceased: "ON", birthday: "2018-2-2" }, { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should validate null property if required", () => {
            const result = validate({ id: null, name: "Mimi", deceased: "ON", birthday: "2018-2-2" }, { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should validate empty string property", () => {
            const result = validate({ id: "", name: "", deceased: "ON", birthday: "2018-2-2" }, { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should validate required validation if property not provided", () => {
            const result = validate({ name: "Mimi", deceased: "ON", birthday: "2018-2-2" }, { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should valid if provided null if not required", () => {
            const result = validate({ id: "123", name: null, deceased: null, birthday: null }, { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should valid if provided undefined if not required", () => {
            const result = validate({ id: "123" }, { ...option })
            expect(result).toMatchSnapshot()
        })
    })

    describe("Array of Primitive Type Validation", () => {

        it("Should validate undefined if required", () => {
            const result = validate(["1", undefined, "3"], {
                path: "data",
                type: [Number],
                decorators: [required]
            })
            expect(result).toMatchSnapshot()
        })
        it("Should validate null if required", () => {
            const result = validate(["1", null, "3"], {
                path: "data",
                type: [Number],
                decorators: [required]
            })
            expect(result).toMatchSnapshot()
        })
        it("Should validate empty string if required", () => {
            const result = validate(["1", "", "3"], {
                path: "data",
                type: [Number],
                decorators: [required]
            })
            expect(result).toMatchSnapshot()
        })
        it("Should valid if not required", () => {
            const result = validate(["1", null, "3"], {
                path: "data",
                type: [Number],
            })
            expect(result).toMatchSnapshot()
        })
    })

    describe("Array of Object Validator", () => {
        @reflect.parameterProperties()
        class AnimalClass {
            constructor(
                @val.required()
                public id: number,
                public name: string,
                public deceased: boolean,
                public birthday: Date
            ) { }
        }

        const option = { type: [AnimalClass], }

        it("Should validate undefined property if required", () => {
            const result = validate([undefined, { id: undefined, name: "Mimi", deceased: "ON", birthday: "2018-2-2" }], { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should validate null property if required", () => {
            const result = validate([undefined, { id: null, name: "Mimi", deceased: "ON", birthday: "2018-2-2" }], { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should validate empty string property if required", () => {
            const result = validate([undefined, { id: "", name: "Mimi", deceased: "ON", birthday: "2018-2-2" }], { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should valid if not required", () => {
            const result = validate([undefined, { id: "123" }], { ...option })
            expect(result).toMatchSnapshot()
        })
    })

    describe("Nested Object Validator", () => {
        @reflect.parameterProperties()
        class TagClass {
            constructor(
                @val.required()
                public id: number,
                public name: string
            ) { }
        }

        @reflect.parameterProperties()
        class AnimalClass {
            constructor(
                public id: number,
                public name: string,
                public deceased: boolean,
                public birthday: Date,
                public tag: TagClass
            ) { }
        }

        const option = { type: AnimalClass, }

        it("Should validate undefined property if required", () => {
            const result = validate({ id: "123", name: "Mimi", deceased: "ON", birthday: "2018-2-2", tag: { id: undefined, name: "The Tag" } }, { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should validate null property if required", () => {
            const result = validate({ id: "123", name: "Mimi", deceased: "ON", birthday: "2018-2-2", tag: { id: null, name: "The Tag" } }, { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should validate empty string property if required", () => {
            const result = validate({ id: "123", name: "Mimi", deceased: "ON", birthday: "2018-2-2", tag: { id: "", name: "The Tag" } }, { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should valid if provided null not required", () => {
            const result = validate({ id: "123", name: "Mimi", deceased: "ON", birthday: "2018-2-2", tag: { id: "123", name: null } }, { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should valid if provided undefined not required", () => {
            const result = validate({ id: "123", tag: { id: "123", name: undefined } }, { ...option })
            expect(result).toMatchSnapshot()
        })
    })

    describe("Partial Validation", () => {

        it("Should skip required validator on partial parent", () => {
            @reflect.parameterProperties()
            class AnimalClass {
                constructor(
                    public id: number,
                    public name: string,
                    @val.required()
                    public deceased: boolean,
                    public birthday: Date
                ) { }
            }
            const result = validate({ id: "123" }, {
                decorators: [<ValidatorDecorator>{ type: "tc:validator", validator: PartialValidator }],
                path: "data",
                type: AnimalClass,
            })
            expect(result).toMatchSnapshot()
        })

        it("Should able to combined with required", () => {
            @reflect.parameterProperties()
            class AnimalClass {
                constructor(
                    public id: number,
                    public name: string,
                    public deceased: boolean,
                    public birthday: Date
                ) { }
            }
            const result = validate(undefined, {
                decorators: [<ValidatorDecorator>{ type: "tc:validator", validator: PartialValidator }, required],
                path: "data",
                type: AnimalClass,
            })
            expect(result).toMatchSnapshot()
        })
    })

})

describe("Validator", () => {
    it("Should validate properly", () => {
        @reflect.parameterProperties()
        class Dummy {
            constructor(
                @val.email()
                public property: string
            ) { }
        }
        const result = validate({ property: "lorem.ipsum@gmail.com" }, { type: Dummy, })
        expect(result).toMatchObject({ value: { property: "lorem.ipsum@gmail.com" } })
    })


    it("Should return message if provided invalid value", () => {
        @reflect.parameterProperties()
        class Dummy {
            constructor(
                @val.email()
                public property: string
            ) { }
        }
        const result = validate({ property: "lorem ipsum" }, { type: Dummy, })
        expect(result).toMatchObject({ issues: [{ messages: ["Invalid email address"], path: "property" }] })
    })

    it("Should prioritize Required validator than other validator", () => {
        @reflect.parameterProperties()
        class Dummy {
            constructor(
                @val.required()
                @val.email()
                public property: string
            ) { }
        }
        const result = validate({ property: "" }, { type: Dummy, })
        expect(result.issues).toMatchObject([{ path: "property", messages: ["Required"] }])
    })

    it("Should valid if not required", () => {
        @reflect.parameterProperties()
        class Dummy {
            constructor(
                @val.email()
                public property: string
            ) { }
        }
        const result = validate({ property: undefined }, { type: Dummy, })
        expect(result).toMatchObject({ value: {}, issues: undefined })
    })

    it("Should able to add custom visitor", () => {
        const visitor = (i: VisitorInvocation) => {
            if (i.type === String) return Result.create("lorem ipsum")
            else return i.proceed()
        }
        @reflect.parameterProperties()
        class Dummy {
            constructor(
                @val.email()
                public property: string
            ) { }
        }
        const result = validate({ property: "lorem.ipsum@gmail.com" }, { visitors: [visitor], type: Dummy, })
        expect(result).toMatchObject({ value: { property: "lorem ipsum" } })
    })
})
