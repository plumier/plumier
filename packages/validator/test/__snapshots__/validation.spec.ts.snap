// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Required Validator and Partial Array of Object Validator Should valid if not required 1`] = `
Object {
  "issues": undefined,
  "value": Array [
    undefined,
    AnimalClass {
      "id": 123,
    },
  ],
}
`;

exports[`Required Validator and Partial Array of Object Validator Should validate empty string property if required 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "1.id",
    },
  ],
  "value": Array [
    undefined,
    AnimalClass {
      "birthday": 2018-1-2,
      "deceased": true,
      "id": "",
      "name": "Mimi",
    },
  ],
}
`;

exports[`Required Validator and Partial Array of Object Validator Should validate null property if required 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "1.id",
    },
  ],
  "value": Array [
    undefined,
    AnimalClass {
      "birthday": 2018-1-2,
      "deceased": true,
      "id": null,
      "name": "Mimi",
    },
  ],
}
`;

exports[`Required Validator and Partial Array of Object Validator Should validate undefined property if required 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "1.id",
    },
  ],
  "value": Array [
    undefined,
    AnimalClass {
      "birthday": 2018-1-2,
      "deceased": true,
      "id": undefined,
      "name": "Mimi",
    },
  ],
}
`;

exports[`Required Validator and Partial Array of Primitive Type Validation Should valid if not required 1`] = `
Object {
  "issues": undefined,
  "value": Array [
    1,
    null,
    3,
  ],
}
`;

exports[`Required Validator and Partial Array of Primitive Type Validation Should validate empty string if required 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "data.1",
    },
  ],
  "value": Array [
    1,
    "",
    3,
  ],
}
`;

exports[`Required Validator and Partial Array of Primitive Type Validation Should validate null if required 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "data.1",
    },
  ],
  "value": Array [
    1,
    null,
    3,
  ],
}
`;

exports[`Required Validator and Partial Array of Primitive Type Validation Should validate undefined if required 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "data.1",
    },
  ],
  "value": Array [
    1,
    undefined,
    3,
  ],
}
`;

exports[`Required Validator and Partial Decorators Should able to decorate partial validator 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [
    Object {
      "decorators": Array [
        Object {
          Symbol(tinspector:decoratorOption): Object {
            "allowMultiple": true,
            "applyTo": Array [],
            "inherit": true,
            "removeApplied": true,
          },
        },
      ],
      "kind": "Method",
      "name": "save",
      "parameters": Array [
        Object {
          "decorators": Array [
            Object {
              "genericParams": Array [],
              "kind": "Override",
              "target": AnimalController,
              "type": AnimalClass,
              Symbol(tinspector:decoratorId): Symbol(override),
              Symbol(tinspector:decoratorOption): Object {
                "allowMultiple": false,
                "applyTo": Array [],
                "inherit": true,
                "removeApplied": true,
              },
            },
            Object {
              "type": "tc:validator",
              "validator": Symbol(tc:partial),
              Symbol(tinspector:decoratorOption): Object {
                "allowMultiple": true,
                "applyTo": Array [],
                "inherit": true,
                "removeApplied": true,
              },
            },
          ],
          "fields": Object {},
          "index": 0,
          "kind": "Parameter",
          "name": "data",
          "type": AnimalClass,
          "typeClassification": "Class",
        },
      ],
      "returnType": undefined,
      "typeClassification": undefined,
    },
  ],
  "name": "AnimalController",
  "properties": Array [],
  "super": Object,
  "type": AnimalController,
  "typeClassification": "Class",
}
`;

exports[`Required Validator and Partial Decorators Should able to decorate required validator 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [
      Object {
        "decorators": Array [],
        "fields": Object {},
        "index": 0,
        "kind": "Parameter",
        "name": "id",
        "type": Number,
        "typeClassification": "Primitive",
      },
      Object {
        "decorators": Array [],
        "fields": Object {},
        "index": 1,
        "kind": "Parameter",
        "name": "name",
        "type": String,
        "typeClassification": "Primitive",
      },
      Object {
        "decorators": Array [],
        "fields": Object {},
        "index": 2,
        "kind": "Parameter",
        "name": "deceased",
        "type": Boolean,
        "typeClassification": "Primitive",
      },
      Object {
        "decorators": Array [
          Object {
            "type": "tc:validator",
            "validator": Symbol(tc:required),
            Symbol(tinspector:decoratorOption): Object {
              "allowMultiple": true,
              "applyTo": Array [],
              "inherit": true,
              "removeApplied": true,
            },
          },
        ],
        "fields": Object {},
        "index": 3,
        "kind": "Parameter",
        "name": "birthday",
        "type": Date,
        "typeClassification": "Primitive",
      },
    ],
  },
  "decorators": Array [
    Object {
      "type": "ParameterProperties",
      Symbol(tinspector:decoratorId): Symbol(paramProp),
      Symbol(tinspector:decoratorOption): Object {
        "allowMultiple": false,
        "applyTo": Array [],
        "inherit": true,
        "removeApplied": true,
      },
    },
  ],
  "kind": "Class",
  "methods": Array [],
  "name": "AnimalClass",
  "properties": Array [
    Object {
      "decorators": Array [],
      "get": undefined,
      "index": 0,
      "isParameter": true,
      "kind": "Property",
      "name": "id",
      "set": undefined,
      "type": Number,
      "typeClassification": "Primitive",
    },
    Object {
      "decorators": Array [],
      "get": undefined,
      "index": 1,
      "isParameter": true,
      "kind": "Property",
      "name": "name",
      "set": undefined,
      "type": String,
      "typeClassification": "Primitive",
    },
    Object {
      "decorators": Array [],
      "get": undefined,
      "index": 2,
      "isParameter": true,
      "kind": "Property",
      "name": "deceased",
      "set": undefined,
      "type": Boolean,
      "typeClassification": "Primitive",
    },
    Object {
      "decorators": Array [
        Object {
          "type": "tc:validator",
          "validator": Symbol(tc:required),
          Symbol(tinspector:decoratorOption): Object {
            "allowMultiple": true,
            "applyTo": Array [],
            "inherit": true,
            "removeApplied": true,
          },
        },
      ],
      "get": undefined,
      "index": 3,
      "isParameter": true,
      "kind": "Property",
      "name": "birthday",
      "set": undefined,
      "type": Date,
      "typeClassification": "Primitive",
    },
  ],
  "super": Object,
  "type": AnimalClass,
  "typeClassification": "Class",
}
`;

exports[`Required Validator and Partial Nested Object Validator Should valid if provided null not required 1`] = `
Object {
  "issues": undefined,
  "value": AnimalClass {
    "birthday": 2018-1-2,
    "deceased": true,
    "id": 123,
    "name": "Mimi",
    "tag": TagClass {
      "id": 123,
      "name": null,
    },
  },
}
`;

exports[`Required Validator and Partial Nested Object Validator Should valid if provided undefined not required 1`] = `
Object {
  "issues": undefined,
  "value": AnimalClass {
    "id": 123,
    "tag": TagClass {
      "id": 123,
      "name": undefined,
    },
  },
}
`;

exports[`Required Validator and Partial Nested Object Validator Should validate empty string property if required 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "tag.id",
    },
  ],
  "value": AnimalClass {
    "birthday": 2018-1-2,
    "deceased": true,
    "id": 123,
    "name": "Mimi",
    "tag": TagClass {
      "id": "",
      "name": "The Tag",
    },
  },
}
`;

exports[`Required Validator and Partial Nested Object Validator Should validate null property if required 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "tag.id",
    },
  ],
  "value": AnimalClass {
    "birthday": 2018-1-2,
    "deceased": true,
    "id": 123,
    "name": "Mimi",
    "tag": TagClass {
      "id": null,
      "name": "The Tag",
    },
  },
}
`;

exports[`Required Validator and Partial Nested Object Validator Should validate undefined property if required 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "tag.id",
    },
  ],
  "value": AnimalClass {
    "birthday": 2018-1-2,
    "deceased": true,
    "id": 123,
    "name": "Mimi",
    "tag": TagClass {
      "id": undefined,
      "name": "The Tag",
    },
  },
}
`;

exports[`Required Validator and Partial Object Validator Should valid if provided null if not required 1`] = `
Object {
  "issues": undefined,
  "value": AnimalClass {
    "birthday": null,
    "deceased": null,
    "id": 123,
    "name": null,
  },
}
`;

exports[`Required Validator and Partial Object Validator Should valid if provided undefined if not required 1`] = `
Object {
  "issues": undefined,
  "value": AnimalClass {
    "id": 123,
  },
}
`;

exports[`Required Validator and Partial Object Validator Should validate empty string property 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "id",
    },
  ],
  "value": AnimalClass {
    "birthday": 2018-1-2,
    "deceased": true,
    "id": "",
    "name": "",
  },
}
`;

exports[`Required Validator and Partial Object Validator Should validate null property if required 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "id",
    },
  ],
  "value": AnimalClass {
    "birthday": 2018-1-2,
    "deceased": true,
    "id": null,
    "name": "Mimi",
  },
}
`;

exports[`Required Validator and Partial Object Validator Should validate required validation if property not provided 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "id",
    },
  ],
  "value": AnimalClass {
    "birthday": 2018-1-2,
    "deceased": true,
    "name": "Mimi",
  },
}
`;

exports[`Required Validator and Partial Object Validator Should validate undefined property if required 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "id",
    },
  ],
  "value": AnimalClass {
    "birthday": 2018-1-2,
    "deceased": true,
    "id": undefined,
    "name": "Mimi",
  },
}
`;

exports[`Required Validator and Partial Partial Validation Should able to combined with required 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "data",
    },
  ],
  "value": undefined,
}
`;

exports[`Required Validator and Partial Partial Validation Should skip required validator on partial parent 1`] = `
Object {
  "issues": undefined,
  "value": AnimalClass {
    "id": 123,
  },
}
`;

exports[`Required Validator and Partial Primitive Type Validation Should give proper info if provided wrong type of value 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Unable to convert \\"hula\\" into Number",
      ],
      "path": "data",
    },
  ],
  "value": "hula",
}
`;

exports[`Required Validator and Partial Primitive Type Validation Should valid if not required and provided empty string 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Unable to convert \\"\\" into Number",
      ],
      "path": "data",
    },
  ],
  "value": "",
}
`;

exports[`Required Validator and Partial Primitive Type Validation Should valid if not required and provided null 1`] = `
Object {
  "value": null,
}
`;

exports[`Required Validator and Partial Primitive Type Validation Should valid if not required and provided undefined 1`] = `
Object {
  "value": undefined,
}
`;

exports[`Required Validator and Partial Primitive Type Validation Should validate empty string if required 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "data",
    },
  ],
  "value": "",
}
`;

exports[`Required Validator and Partial Primitive Type Validation Should validate null if required 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "data",
    },
  ],
  "value": null,
}
`;

exports[`Required Validator and Partial Primitive Type Validation Should validate undefined if required 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "data",
    },
  ],
  "value": undefined,
}
`;
